### ISSUE TITLE
cannot copy unicode characters

### ISSUE THREAD

**Issue by haifengkao** [neovim/neovim#5683]
- `nvim --version`: NVIM 0.2.0-dev (9956bee)
- Vim (version: ) behaves differently?  Yes
- Operating system/version: OSX 10.11
- Terminal name/version: iTerm
- `$TERM`: xterm-256color

### Actual behaviour
The unicode characters like `中文` cannot be copied.

### Expected behaviour
It should be copied to registers.

### Steps to reproduce using `nvim -u NORC`

1. Open `System Preference` -> `Language & Region` in OSX
2. Choose a non-English language as the preferred language (I use Chinese)
3. Restart your mac
4. Install a standalone Nyaovim from here https://github.com/haifengkao/NyaoVim-Unofficial/releases
5. double click the Nyaovim.app
6. type `中文<ESC>yyp`, it will not be copied

This issue is specified to Nyaovim, see https://github.com/rhysd/NyaoVim/issues/77
However, it is NOT caused by Nyaovim.

If Nyaovim.app is opened from terminal with `open /Applications/Nyaovim.app`, it will copy the chars correctly.

 I think it is related to the locale detection of `neovim`. For some reason, the locale is not set correctly when it is not activated from the terminal and the system's preferred language is NOT English.

**Comment by justinmk** [comment]
> Install a standalone Nyaovim from here 

We need steps with `nvim -u NORC`.

> I think it is related to the locale detection of neovim

What does `:lang` say?

**Comment by haifengkao** [comment]
I don't know how to start nyaovim with `nvim -u NORC`
If I remove everything in `~/.config`, `中文<ESC>yyp` will work correctly in Nyaovim.

**Comment by justinmk** [comment]
> If I remove everything in `~/.config`, `中文<ESC>yyp` will work correctly in Nyaovim.

Can you narrow down to the specific option that triggers the problem?

**Comment by haifengkao** [comment]
If I set clipboard to non-empty, unicode chars cannot be copied.
e.g.
`set clipboard=unnamed`
or
`set clipboard=unnamedplus`

**Comment by justinmk** [comment]
Cannot reproduce. These steps work correctly, for me:

```
nvim -u NORC
:set clipboard=unnamed
中文<ESC>yyp
```

Also works with `"+p`, cmd+v, and `<C-R>+`.

**Comment by justinmk** [comment]
> This issue is specified to Nyaovim, see rhysd/NyaoVim#77
> However, it is NOT caused by Nyaovim.

Why?

> If Nyaovim.app is opened from terminal with open /Applications/Nyaovim.app, it will copy the chars correctly.
> I think it is related to the locale detection of neovim. 

It works in normal `nvim` (as in my steps above), and it works when nyaovim.app is "opened from terminal with `open /Applications/Nyaovim.app`". What makes you think `nvim` is root cause of the issue?

> For some reason, the locale is not set correctly when it is not activated from the terminal and the system's preferred language is NOT English.

Can you mention what `:lang` say, in both cases?

**Comment by haifengkao** [comment]
Yes, it works for `nvim`.

The `:lang` for `nvim` 
```
Current language:›"en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
```
The `:lang` for standalone `nyaovim`
![2016-11-29 3 35 50](https://cloud.githubusercontent.com/assets/4080524/20682956/24921a1a-b5e5-11e6-882d-f429822de2a2.jpg)

The `:lang` for `open /Applications/Nyaovim.app`
![2016-11-29 3 40 58](https://cloud.githubusercontent.com/assets/4080524/20683134/c559f238-b5e5-11e6-8306-99659f32768d.jpg)


> What makes you think nvim is root cause of the issue?

Because this issue can be simply reproduced by 
```javascript
document.getElementById('nyaovim-editor').editor.getClient().input('<C-c>')
```
It means nyaovim doesn't do anything other than sending the input directly to `neovim`.
I can put unicode stuffs in registers except `+` or `*`.  It means the encoding between `nyaovim` and `neovim` is consistent, but it is inconsistent between`neovim` and `pbcopy`.

**Comment by justinmk** [comment]
Perhaps related to `iconv`. Is `zh_TW` a single-byte encoding? In NyaoVim, try `:lang zh_TW.UTF-8` (just a guess).

If you run `nvim` from terminal, then change the language does the problem occur?:

    :lang zh_TW

I observe a similar issue with these steps:

```
nvim
" insert 中文
:set clipboard=unnamed
:lang en_AU.US-ASCII
yy
```

After the `yy` step, clipboard contains the wrong contents:

    ‰∏≠Êñá

**Comment by haifengkao** [comment]
> Is zh_TW a single-byte encoding?

No, zh_TW is CNS-11643, so it is 2-byte encoding (https://github.com/macvim-dev/macvim/blob/master/runtime/doc/mbyte.txt).

>In NyaoVim, try :lang zh_TW.UTF-8 (just a guess).

Yes, `中文<ESC>yyp` works without any problems.

>If you run nvim from terminal, then change the language does the problem occur?:

No, `:lang zh_TW` still works in nvim without any problems.

Below is the output of `:lang` from `MacVim`. It seems that `zh_TW` is indeed the system's encoding.
![2016-11-29 10 35 51](https://cloud.githubusercontent.com/assets/4080524/20694367/a17a7380-b61f-11e6-9bba-092f46575ef8.jpg)

**Comment by justinmk** [comment]
>> In NyaoVim, try :lang zh_TW.UTF-8 (just a guess).

> Yes, `中文<ESC>yyp` works without any problems.

@haifengkao So you have a workaround for now?

**Comment by haifengkao** [comment]
Yes, just put ` :lang zh_TW.UTF-8` in `init.vim`

**Comment by thinker3** [comment]
@justinmk 
very sad, so many days. I need to select the file path, mostly by double click, then copy it

**Comment by justinmk** [comment]
@thinker3 Did you try the work around?

**Comment by justinmk** [comment]
Reference: https://github.com/vim/vim/issues/1082

**Comment by fturcheti** [comment]
I had a similar problem: when copying texts, letters with accents were swapped by other characters (eg. copying `são` would put `s√£o` in the clipboard). This was happening only in _terminal Neovim_. I tried it in _VimR_, _Oni_, _Vim_ and _MacVim_ and it was working properly everywhere.

After finding this Github issue, I ran `:lang` as @justinmk suggested and the result was: `Current language: "C"`.

So I followed the workaround by @haifengkao, added `lang en_US.UTF-8` to my `init.vim`, and the problem was fixed.

**Comment by cseelus** [comment]
Had the same problem, It appeared with NeoVIM 0.3.2 I guess. Adding `lang en_US.UTF-8` helps, but UTF-8 should be supported by default nowadays.

**Comment by justinmk** [comment]
Perhaps https://github.com/neovim/neovim/pull/9789 fixed this. Can you try current HEAD?  

The [Releases](https://github.com/neovim/neovim/releases) page has pre-built archives for Linux/Windows/macOS. (No installation needed, just unzip and run. For Linux, [nvim.appimage](https://github.com/neovim/neovim/releases/download/nightly/nvim.appimage) is a universal "binary", just `chmod u+x nvim.appimage && ./nvim.appimage`)

**Comment by fturcheti** [comment]
In this pre-realease version (NVIM v0.4.0-603-gd81b510ec) `:lang` returns `Current language: "C/UTF-8/C/C/C/C"`, not just `Current language: "C"` as in NVIM v0.3.4, but the reported problem still occurs.

Similar to the latest Neovim pre-release, when I run `:lang` in Vim (VIM 8.1; Included patches: 1-1150) I get the same `Current language: "C/UTF-8/C/C/C/C"`, although I don't see this problem occurring there.

**Comment by justinmk** [comment]
Is this still an issue with latest 0.4 [prelrease](https://github.com/neovim/neovim/releases) ? I am wondering if https://github.com/neovim/neovim/pull/10697 fixed this.

**Comment by cseelus** [comment]
@justinmk I've removed `lang en_US.UTF-8` from my `init.vim` and it seems this is fixed now with the latest prerelease.

**Comment by MMDF** [comment]
This is still an issue in 0.4.4 and only fixed by setting lang to en_US.UTF-8 for me on macOS. (default lang was C)

**Comment by kriansa** [comment]
@justinmk I can confirm this bug is back.

**Comment by aignas** [comment]
I am running `alacritty + tmux + nvim 0.5` on MacOS and I can confirm that the issue is solved if I set:
```
:lang en_US.UTF-8
```

I am using the official `brew` package of neovim. Is the default behaviour expected? @justinmk, should we open a new github issue ticket for this instead?

**Comment by Liverm0r** [comment]
has this bug by default in any vim client on osx (vimr, neovim in default terminal, neovide), only `lang en_US.UTF-8` helps.

**Comment by clason** [comment]
please try the latest master/nightly; we should now always set a sane locale.

(This fix will be in 0.7.1.)
